ikarus/utils/tensorutils.hh

```diff
@@ -219,19 +219,33 @@ Eigen::Matrix<ScalarType, 6, 6> toVoigt(const Eigen::TensorFixedSize<ScalarType,
  * calculations. If isStrain is true, the off-diagonal components are multiplied by 2, providing the correct Voigt
  * notation for symmetric strain tensors.
  */
-template <typename ST, int size, int Options>
-requires(size > 0 and size <= 3)
-auto toVoigt(const Eigen::Matrix<ST, size, size, Options, size, size>& E, bool isStrain = true) {
-  Eigen::Vector<ST, (size * (size + 1)) / 2> EVoigt;
-  EVoigt.template segment<size>(0) = E.diagonal();
 
+template <typename ST, int size, int Options, int maxSize>
+requires((size > 0 and size <= 3) or (maxSize > 0 and maxSize <= 3 and size == Eigen::Dynamic))
+auto toVoigt(const Eigen::Matrix<ST, size, size, Options, maxSize, maxSize>& E, bool isStrain = true) {
+  constexpr bool isFixedSized   = (size != Eigen::Dynamic);
   const ST possibleStrainFactor = isStrain ? 2.0 : 1.0;
-  if constexpr (size == 2)
+
+  const size_t inputSize = isFixedSized ? size : E.rows();
+  auto EVoigt            = [&]() {
+    if constexpr (isFixedSized) {
+      Eigen::Vector<ST, (size * (size + 1)) / 2> EVoigt;
+      EVoigt.template segment<size>(0) = E.diagonal();
+      return EVoigt;
+    } else {
+      Eigen::VectorX<ST> EVoigt;
+      EVoigt.resize((inputSize * (inputSize + 1)) / 2);
+      EVoigt.template head<Eigen::Dynamic>(inputSize) = E.diagonal();
```

Comment:
Why `Eigen::Dynamic` ? and maybe you can also use head above?

I think you can just write  EVoigt.template head<size>(inputSize) for the dynamic and static case.